cypress-io/cypress
# fix: improve memory management for chromium-based browsers

- Closes #23391

### User facing changelog

* Added new `experimentalMemoryManagement` flag to improve memory management in Chromium-based browsers.

### Additional details

Algorithm:
 
 * When the test run starts:
    1. set total mem limit for the container/host by reading off cgroup memory limits (if available) otherwise use os.totalmem()
    2. turn on memory profiler
 
 * On a defined interval (e.g. 1s):
    1. set current mem available for the container/host by reading off cgroup memory usage (if available) otherwise use si.mem().available
    3. set current renderer mem usage
    4. set max avail render mem to minimum of v8 heap size limit and total available mem (current available mem + current renderer mem usage)
    5. calc % of memory used, current renderer mem usage / max avail render mem
 
 * Before each test:
    1. if any interval exceeded the defined memory threshold, then do a GC prior in between running the next test
 * After the test run ends:
     1. turn off memory profiler

Since the manual GC is run before each test, if a single test consumes enough memory, it is still possible to reach an OOM scenario.

### Steps to test

Run a spec that consumes enough memory to cause an OOM error and then turn on the experimental flag and verify the spec now passes.

### How has the user experience changed?

Ran the following spec on Chromium 107 in run mode:

```
describe('memory', () => {
  for (let index = 0; index < 200; index++) {
    it(`test ${index + 1}`, () => {
      cy.visit('https://earth.google.com/web/')
    })
  }
})
```

Before (OOM crash at test 85):
<img width="1393" alt="Screen Shot 2023-01-13 at 12 05 42 PM" src="https://user-images.githubusercontent.com/2002044/212398936-698002f3-65fd-4c5e-9d69-18453f095c0a.png">

After (No OOM crash):
<img width="1327" alt="Screen Shot 2023-01-13 at 12 03 06 PM" src="https://user-images.githubusercontent.com/2002044/212398545-40982913-eb52-48bd-8c5d-94252fae01a7.png">

### PR Tasks
